### DISCUSSION TITLE
Best way to maintain the order of a dict (e.g. `fromJson`)

### DISCUSSION THREAD

**Discussion by bradenhilton** [twpayne/chezmoi#2864] [Q&A]
I'm trying to manage Scoop's config with chezmoi using the following `modify_` template:

```text
// chezmoi:modify-template
// chezmoi:template:line-endings=crlf

{{- .chezmoi.stdin | fromJson |
    setValueAtPath "aria2-enabled" true |
    setValueAtPath "aria2-warning-enabled" false |
    setValueAtPath "use_external_7zip" true |
    toPrettyJson -}}
```

This outputs the keys in the `dict` as JSON properties in alphabetical order:

```text
{
  "aria2-enabled": true,
  "aria2-warning-enabled": false,
  "last_update": "2023-03-16T09:36:35.1697245+00:00",
  "scoop_repo": "https://github.com/ScoopInstaller/Scoop",
  "scoop_branch": "master",
  "use_external_7zip": true
}
```

This works until Scoop updates its own config (notice the `last_update` property), where it moves `use_external_7zip` (and possibly even more properties, depending on which ones are set):

```text
{
  "aria2-enabled": true,
  "aria2-warning-enabled": false,
  "last_update": "2023-03-16T11:22:57.3547108+00:00",
  "use_external_7zip": true,
  "scoop_repo": "https://github.com/ScoopInstaller/Scoop",
  "scoop_branch": "master"
}
```

This obviously results in unwanted diffs.

Currently I'm resorting to constructing the JSON manually:

```text
// chezmoi:modify-template
// chezmoi:template:line-endings=crlf

{{- $config := .chezmoi.stdin | fromJson -}}
{{- $_ := set $config "aria2-enabled" true -}}
{{- $_ := set $config "aria2-warning-enabled" false -}}
{{- $_ := set $config "use_external_7zip" true -}}

{{- /* Construct the JSON manually to preserve its order */ -}}
{
  "aria2-enabled": {{ get $config "aria2-enabled" | toJson -}},
  "aria2-warning-enabled": {{ get $config "aria2-warning-enabled" | toJson -}},
  "last_update": {{ get $config "last_update" | toJson -}},
  "use_external_7zip": {{ get $config "use_external_7zip" | toJson -}},
  "scoop_repo": {{ get $config "scoop_repo" | toJson -}},
  "scoop_branch": {{ get $config "scoop_branch" | toJson }}
}
```

(There is probably a nicer way to do this manual construction, I'm just experimenting at the moment)

This approach prevents the diffs, but it now means I need to keep track of all properties in Scoop's config and manually add/re-order them. I could also potentially edit the raw text of the config, which would preserve its order, but comes with its own problems.

I am aware that `dict`s are unordered, and that JSON is _not supposed_ to be ordered. But in this niche case, I need it to be.

Have I missed anything that might help? Could we even potentially add ordered `dict` support for this kind of scenario?

**Comment by halostatue** [comment]
This isn’t even a meaningfully ordered dict by scoop, as it appears to always put `scoop_*` last. You could build some gnarly template logic that looks something like this:

```
// chezmoi:modify-template
// chezmoi:template:line-endings=crlf
{{- $firstKeys := ["aria2-enabled", "aria2-warning-enabled"] -}}
{{- $lastKeys := ["scoop_repo", "scoop_branch"] -}}
{{- $config := .chezmoi.stdin | fromJson -}}
{{- $_ := set $config "aria2-enabled" true -}}
{{- $_ := set $config "aria2-warning-enabled" false -}}
{{- $_ := set $config "use_external_7zip" true -}}
{
  {{ range $firstKeys }}{{ . | toJson }}: {{ get $config . | toJson -}},
  {{ end -}}
  {{- range $k, $v := $config }}{{ if not (or (has $k $firstKeys) (has $k $lastKeys)) }}{{ $k | toJson }}: {{ $v | toJson -}},
  {{ end }}{{ end -}}
  {{- range $i, $e := $lastKeys }}{{ $e | toJson }}: {{ get $config $e | toJson -}}
    {{- if ne $i (len $lastKeys) }},{{- end -}}
  {{- end }}
}
```

Not actually tested, and really ugly, but it lets you choose certain keys to fix in certain positions.

**Reply by bradenhilton** [reply]
I can grab the keys in their existing order with regex plus some trimming, which is arguably even uglier:

```text
# I'd rather use a named capture group here to remove the need to loop/trim/append to a list,
# but regexFindAll just returns a string slice
{{- $keyRegex := `"[^"]+":` -}}

{{- $keys := list -}}
{{- range regexFindAll $keyRegex .chezmoi.stdin -1 -}}
{{-   $key := trimPrefix `"` (trimSuffix `":` .) -}}
{{-   $keys = append $keys $key -}}
{{- end -}}
```

But it does somewhat simplify the JSON construction:

```text
{
{{ range $index, $key := $keys -}}
{{-  indent 2 (printf "%v: %v" ($key | toJson) (get $config $key | toJson)) -}}{{- if ne (add1 $index) (len $config) -}}{{- print ",\r\n" -}}{{- end -}}
{{- end }}
}
```

I'm sure it would also be possible to grab the indentation and key/value separator sizes from the input text and use those if needed as well.

Imperfect, but seems like a reasonable approach without changing any code. Any thoughts?

**Reply by twpayne** [reply]
I am simultaneously extremely impressed and extremely horrified. It's a hack (in the original sense of the word - making something do a thing that the something was never designed to do). But if it works then awesome.

**Comment by twpayne** [comment]
Note that `toJson` emits keys alphabetically. It's Scoop that's changing the order of keys here.

Adding an ordered dict is a lot of work to implement:

1. It would require a specific ordered dict type.
2. It would require a new JSON parser that uses the ordered dict type, as Go's standard library uses `map[string]any`, which is unordered.
3. All template functions that handle dicts would have to be modified to also accept ordered dict. 

So, I would report this as a bug in Scoop and ask them to either preserve dict order or always emit keys in order.

**Reply by bradenhilton** [reply]
I found the bug in Scoop. Whether or not it will be fixed remains to be seen, so I'll have to keep experimenting.

**Reply by halostatue** [reply]
@twpayne BTW, I think that @bradenhilton has found the *first case* where the `jq` filter would be useful. Maybe (I say maybe because gojq may always sort the keys on output regardless of any manipulation; with *actual* `jq`, it would be easier).

**Reply by bradenhilton** [reply]
@halostatue `gojq` would be unsuitable for this use case, no?

> - gojq does not keep the order of object keys. I understand this might cause problems for some scripts but basically, we should not rely on the order of object keys. Due to this limitation, gojq does not have `keys_unsorted` function and `--sort-keys` (`-S`) option. I would implement when ordered map is implemented in the standard library of Go but I'm less motivated.

**Reply by halostatue** [reply]
As far as I can tell, `gojq` *always* sorts the keys.

```console
$ echo '{"aria2-enabled":true,"aria2-warning-enabled":false,"last_update":"2023-03-16T11:22:57.3547108+00:00","use_external_7zip":true,"scoop_repo":"https://github.com/ScoopInstaller/Scoop","scoop_branch":"master"}' | gojq .
{
  "aria2-enabled": true,
  "aria2-warning-enabled": false,
  "last_update": "2023-03-16T11:22:57.3547108+00:00",
  "scoop_branch": "master",
  "scoop_repo": "https://github.com/ScoopInstaller/Scoop",
  "use_external_7zip": true
}
```

This makes it *unfortunately* unsuitable for your use case as well.

**Comment by bradenhilton** [comment]
It just occurred to me that one could avoid this issue by comparing the string representations of both the current file contents and their modified values. If the strings are equal, you can just opt to output `.chezmoi.stdin`, which would obviously maintain the order.

```text
{{ $currentConfig := .chezmoi.stdin | fromJson }}
{{ $modifiedConfig := .chezmoi.stdin | fromJson }}

{{/* Set values in $modifiedConfig here */}}

{{ if eq ($currentConfig | toJson) ($modifiedConfig | toJson) }}
{{   .chezmoi.stdin }}
{{ else }}
{{   $modifiedConfig | toPrettyJson }}
{{ end }}
```

This works for my Scoop config (`modify_config.json`):

```text
// chezmoi:modify-template
// chezmoi:template:line-endings=crlf
// chezmoi:template:left-delimiter=//{{
//{{- $currentConfig := .chezmoi.stdin | fromJson -}}
//{{- $modifiedConfig := .chezmoi.stdin | fromJson -}}

//{{- $_ := set $modifiedConfig "use_external_7zip" true -}}

//{{- if eq ($currentConfig | toJson) ($modifiedConfig | toJson) -}}
//{{-   .chezmoi.stdin }}
//{{- else -}}
//{{-   $modifiedConfig | toPrettyJson }}
//{{- end -}}
```

I get one initial diff if `use_external_7zip` is not `true`, then no further diffs.

**Reply by twpayne** [reply]
Very clever!

**Reply by bradenhilton** [reply]
I just noticed that Sprig offers `deepEqual`, which should be much more robust than converting to strings and comparing.
